### PR TITLE
Support no token in AnalyticsClient

### PR DESCRIPTION
**Pull request by francoislg**
And some refactoring to pass a custom AnalyticsClient to the simpleanalytics initialization

### CODE REVIEW

File: src/analytics.ts [hunk 1]
```diff
@@ -19,28 +28,25 @@ function defaultResponseTransformer(response: IResponse): Promise<any> {
...
         if (typeof opts === 'undefined') {
             throw new Error('You have to pass options to this constructor');
         }
-        if (typeof opts.token === 'undefined') {
-            throw new Error('You have to pass opts.token');
-        }
 
         this.endpoint = opts.endpoint || Endpoints.default;
-        this.token = opts.token;
+        this.token = opts.token || '';
```
Comment (pastjean) [inline src/analytics.ts]: no need this. let's undefined that

File: src/analytics.ts [hunk 2]
```diff
@@ -49,106 +55,40 @@ export class Client {
...
-    id: string;
-    visitorId: string;
-}
-export interface HealthResponse {
-    raw: IResponse;
-    status: string;
+    protected applyDefaultHeaders(headers: any) {
+        if (this.token !== '') {
+            headers['Authorization'] = `Bearer ${this.token}`;
+            headers['Content-Type'] = `application/json`;
```
Comment (pastjean) [inline src/analytics.ts]: on a toujours besoin de content-type

File: src/analytics.ts [hunk 3]
```diff
@@ -19,28 +28,25 @@ function defaultResponseTransformer(response: IResponse): Promise<any> {
...
 
         this.endpoint = opts.endpoint || Endpoints.default;
-        this.token = opts.token;
+        this.token = opts.token || '';
     }
 
-    sendEvent( eventType: string, request: any): Promise<IResponse> {
+    sendEvent(eventType: string, request: any): Promise<IResponse> {
+        var headers: any = {};
+        this.applyHeaders(headers);
```
Comment (pastjean) [inline src/analytics.ts]: j'aurais probablement laissé tout ça la avec un if qui met ou pas authorization selon si le  token est présent

Comment (pastjean) [inline src/analytics.ts]:
```js
if(this.token){
headers["aut..."]= `Bearer ${this.token}`
}```

File: src/simpleanalytics.ts [hunk 4]
```diff
@@ -1,24 +1,30 @@
...
-            token: token,
-            endpoint: endpoint
-        });
+        if (typeof token === 'string') {
+            endpoint = endpoint || analytics.Endpoints.default;
+            this.client = new analytics.Client({
+                token: token,
+                endpoint: endpoint
+            });
+        } else if (typeof token === 'object' && typeof token.sendEvent !== 'undefined') {
```
Comment (pastjean) [inline src/simpleanalytics.ts]: `typeof token === AnalyticsClient` should work

**Comment by pastjean** [comment]
GG , just fix my comments!

File: src/analytics.ts [hunk 5]
```diff
@@ -8,7 +17,7 @@ export const Endpoints = {
 };
 
 export interface ClientOptions {
-  token: string;
+  token?: string;
```
Comment (wfortin) [inline src/analytics.ts]: Token is optional ?

Comment (wfortin) [inline src/analytics.ts]: Ok I discussed with @pastjean and I understand that you don't need it.

File: src/analytics.ts [hunk 6]
```diff
@@ -19,28 +28,25 @@ function defaultResponseTransformer(response: IResponse): Promise<any> {
...
-    sendEvent( eventType: string, request: any): Promise<IResponse> {
+    sendEvent(eventType: string, request: any): Promise<IResponse> {
+        var headers: any = {};
+        this.applyHeaders(headers);
         return fetch(`${this.endpoint}/analytics/${eventType}`, {
             method: 'POST',
-            headers: {
-                Authorization: `Bearer ${this.token}`,
-                'content-type' : 'application/json'},
+            headers: headers,
```
Comment (wfortin) [inline src/analytics.ts]: You could just call a `getHeaders` method here.

File: src/simpleanalytics.ts (hunk 4 above)
Comment (wfortin) [inline src/simpleanalytics.ts]: This won't work AnalyticsClient is an interface.

File: src/analytics.ts (hunk 3 above)
Comment (lbenie) [inline src/analytics.ts]: I agree with @pastjean

File: test/helpers/analyticsclientmock.ts [hunk 7]
```diff
@@ -0,0 +1,34 @@
+import AnalyticsClient from '../../src/analyticsclient';
+import {
+    SearchEventRequest, SearchEventResponse,
+    ClickEventRequest, ClickEventResponse,
+    CustomEventRequest, CustomEventResponse,
+    ViewEventRequest, ViewEventResponse,
+    VisitResponse, HealthResponse
+} from '../../src/events';
```
Comment (wfortin) [inline test/helpers/analyticsclientmock.ts]: The formatting here is a little weird, we should plug in https://github.com/coveo/pretty-typescript soon!

**Comment by wfortin** [comment]
Not much to add to @pastjean 's comments. Good job.

File: test/helpers/analyticsclientmock.ts (hunk 7 above)
Comment (francoislg) [inline test/helpers/analyticsclientmock.ts]: I preferred this way to the "one-liner to rule them all".

The cool thing is the match between Request & Response on each line.

Although VisitResponse and HealthResponse don't have Request.

Any suggestion ?

File: src/analytics.ts (hunk 3 above)
Comment (francoislg) [inline src/analytics.ts]: Good, changed back to getHeaders() :smile:

File: test/helpers/analyticsclientmock.ts (hunk 7 above)
Comment (pastjean) [inline test/helpers/analyticsclientmock.ts]: import * maybe

Comment (francoislg) [inline test/helpers/analyticsclientmock.ts]: Was my first idea, doesn't  work :/

**Pull request merged**